internal/clock/fake.go

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package clock
+
+// Forked from github.com/andres-erbsen/clock to isolate a missing nap.
+
+import (
+	"container/heap"
+	"runtime"
+	"sync"
+	"time"
+)
+
+// FakeClock represents a fake clock that only moves forward programmically.
+// It can be preferable to a real-time clock when testing time-based functionality.
+type FakeClock struct {
+	sync.Mutex
+
+	now    time.Time
+	timers timers
+}
+
+var _ Clock = (*FakeClock)(nil)
+
+// NewFake returns an instance of a fake clock.
+// The current time of the fake clock on initialization is the Unix epoch.
+func NewFake() *FakeClock {
+	// Note: Unix(0, 0) is not the zero value for time, and we need the zero
+	// value to distinguish rate limiters that have not started.
+	return &FakeClock{now: time.Unix(0, 0)}
```

Comment:
TIL it is not, and the ratelimiter depends on the zero value to detect that it hasn’t seen a real clock value yet. Alternative is to have a different sentinel or separate flag, or initialize with real time.